tensorstore/kvstore/s3/s3_request_builder.h

```diff
@@ -0,0 +1,115 @@
+// Copyright 2020 The TensorStore Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef TENSORSTORE_KVSTORE_S3_REQUEST_BUILDER_H_
+#define TENSORSTORE_KVSTORE_S3_REQUEST_BUILDER_H_
+
+/// \file
+/// S3 Request Builder
+
+#include <string>
+#include <string_view>
+#include <vector>
+
+#include "tensorstore/internal/http/http_request.h"
+#include "tensorstore/internal/path.h"
+#include "tensorstore/kvstore/byte_range.h"
+#include "tensorstore/kvstore/s3/validate.h"
+#include "tensorstore/util/result.h"
+
+#include "absl/time/time.h"
+
+using ::tensorstore::internal_http::HttpRequest;
+using ::tensorstore::internal_http::HttpRequestBuilder;
+
+namespace tensorstore {
+namespace internal_storage_s3 {
+
+std::string UriEncode(std::string_view src);
+std::string UriObjectKeyEncode(std::string_view src);
+
+class S3RequestBuilder {
+ public:
+  explicit S3RequestBuilder(std::string_view method, std::string endpoint_url) :
```

Comment:
explicit not needed.

If this were in the .cc file, then `UriEncode` could be in the inline namespace, if that's desirable.